### PR TITLE
Feature 150 download plot table

### LINKED ISSUE #150: Proof of concept download entire plot table
As a lower-hanging-fruit since we don't have time for a full data-cart-style implementation...
Users would like the ability to download plot “search results” as csv. To do this we'll need:

- Download this table (and download this page?) button
  - Only enabled when table filtered to sufficient extent (20k plots?)
- A utility client-side to fetch the data in the whole table (limit matches count, all comm, all taxa, dist, soils etc)
  - We may need to set a cap on max limit for performance sake (20,000?), only when filtered?
  - Need a loading state as it took 48 seconds to fetch data for all 17,326 plots in SW Gap Nevada proj
- A utility to pull out nested dataframes so they can be made into individual csvs
- A utility that creates those csvs and zips them

### PR DESCRIPTION
**Pull request by DarianGill**
This more detailed PR description is brought to you by Claude, edited for relevancy by @DarianGill 

### What

This PR closes #150 by implementing a data download button for the plot table, allowing users to download up to 20,000 filtered records as a ZIP file containing multiple CSVs with properly structured relational data.

**Key Features:**
- Integrated download button within DataTable using DT Buttons extension
- Automatic extraction of nested data (taxa observations, community classifications) into separate CSV files
- User-friendly CSV filenames, primary key as first column in all CSV files, and structured README documentation
- Real-time download button enable/disable based on filtered record count
- Loading overlay with progress updates during download process generalized from map loading overlay
- Client-side validation to prevent failed download entries in browser history

**Files Added:**
- `R/download_table_data.R` (494 lines) - Core download module
- `tests/testthat/test_download_table_data.R` (391 lines) - 83 tests
- `tests/testthat/test_ui_loading_overlay.R` (239 lines) - 41 tests for loading overlay

**Files Modified:**
- `R/ui.R` - Generalized loading overlay system, client-side validation
- `R/server.R` - Download handler setup, button state management
- `R/table_plot.R` - DT Buttons integration
- `DESCRIPTION` - Added `tidyr` and `zip` dependencies
- `NAMESPACE` - Exported `tidyr` and `zip` imports

### Why

So users can download filtered plot data for offline analysis along with some of the nested structures (taxa observations, community classifications) in a simpler fashion than the previous data cart method.

### How

**Architecture:**

1. **DT Buttons Integration** (`table_plot.R`)
   - Custom button in DataTable triggers download via Shiny input binding
   - Button text clearly indicates 20,000 record limit

2. **Button State Management** (`server.R`)
   - Observer reacts to `plot_table_filtered_count` from DataTable AJAX response
   - Enables button only when: count is known, count > 0, and count ≤ 20,000
   - Uses custom message handler to update button disabled state

3. **Client-Side Validation** (`ui.R`)
   - `triggerDownload` handler validates count before navigating to download URL
   - Prevents browser from recording failed download attempts
   - Shows Shiny notifications for validation failures

4. **Server-Side Download Handler** (`download_table_data.R`)
   - **Configuration-driven**: `TABLE_DOWNLOAD_CONFIG` defines API endpoints, nested structures, and naming for easier extension to other tables in future
   - **Filter state extraction**: Captures search terms and cross-resource filters
   - **Data fetching**: Uses vegbankr to fetch filtered records from API
   - **Nested data detection**: Auto-detects list columns containing data frames
   - **Table extraction**: Uses `tidyr::unnest()` to extract nested tables with foreign keys
   - **CSV preparation**: Removes all list columns from main table, creates separate CSVs for nested data
   - **Column ordering**: Primary key is first column in all tables
   - **ZIP creation**: Uses `zip::zip()` to bundle CSVs with README
   - **Progress updates**: Sends loading overlay updates throughout process
   - **Error handling**: Dismisses overlay and shows notifications on failures

5. **Generalized Loading Overlay** (`ui.R`)
   - `build_loading_overlay()` function creates overlay HTML with data attributes
   - Messages, completion text, and intervals configured via R parameters (not buried in JavaScript)
   - Wrapper functions for specific use cases: `build_map_loading_overlay()`, `build_download_loading_overlay()`
   - JavaScript handlers: `showLoadingOverlay`, `updateLoadingOverlay`, `hideLoadingOverlay`
   - Humorous messages with smooth fade-out animation on completion

**Data Structure:**
vegbank_plots_20260129.zip
├── README.txt (citation, structure explanation, join examples)
├── main.csv (plot data with ob_code as first column)
├── soils.csv (linked by ob_code)
├── disturbances.csv (linked by ob_code)
├── taxon_observations.csv (linked by ob_code)
└── community_classifications.csv (linked by ob_code)

**Key Technical Decisions:**
- Use `tidyr::unnest()` instead of manual nested data extraction for robustness
- Validate on both client and server side for security and UX
- Use client-provided count instead of redundant API call for efficiency
- Map API column names (`top_taxon_observations`) to user-friendly CSV names (`taxon_observations`)
- Simplified ZIP creation using `zip::zip()` directly (it's now a required dependency)

### Testing & Documentation

**Test Coverage:**
- **173 new tests** added across 2 test files
- **Total test suite: 903 tests**, all passing

**Download Module Tests** (`test_download_table_data.R` - 132 tests):
- Filter state extraction (search, cross-resource filters, edge cases)
- Nested column auto-detection (various data types, empty data, mixed columns)
- Table extraction with tidyr::unnest (foreign keys, column ordering, error handling)
- CSV preparation (main table, nested tables, list column removal)
- README generation (filters, dates, structure documentation)
- Integration scenarios (full download flow, error conditions)

**Loading Overlay Tests** (`test_ui_loading_overlay.R` - 41 tests):
- HTML structure and data attributes
- Message encoding with special characters
- ID uniqueness and CSS classes
- Map and download overlay variants
- Edge cases (empty messages, NULL values, special characters)

**Manual Testing Checklist: Chrome, Firefox, & Safari**
- ✅ Download with no filters
- ✅ Download with search term
- ✅ Download with cross-resource filter (e.g., filtered by project)
- ✅ Button disabled when count > 20,000
- ✅ Button disabled when count = 0
- ✅ Loading overlay shows progress updates
- ✅ ZIP contains properly structured CSVs
- ✅ README contains accurate information
- ✅ Foreign keys correctly link related tables
- ✅ Primary key is first column in all CSVs
- ✅ No failed downloads in browser history when validation fails

**Documentation:**
- Comprehensive roxygen comments in `download_table_data.R`
- README.txt included in every download with citation, structure explanation, and join examples
- Clear error messages and notifications guide users when downloads can't proceed
- Button text communicates the 20,000 record limit upfront

**Dependencies Added:**
- `tidyr (>= 1.3.0)` - Robust nested data frame unnesting
- `zip (>= 2.3.2)` - Cross-platform ZIP file creation

### CODE REVIEW

File: R/download_table_data.R [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+    }
+  }
+
+  # Create main table with ALL list columns removed
+  main_df <- data
+  list_cols <- sapply(main_df, is.list)
+
+  if (any(list_cols)) {
+    cols_to_remove <- names(main_df)[list_cols]
+    message("  Removing all list columns from main table: ", paste(cols_to_remove, collapse = ", "))
```
[the remark this answers is not in the document]
Comment (DarianGill) [reply R/download_table_data.R:225]: Will do debug pass later

**Review by regetz — APPROVED** [review]
Thanks @DarianGill -- it works well in my testing, and I like the README!
